File: CarMall_AspBackend/Repository/OrderService.cs
using carManagement.Data;
using carManagement.Interface;
using carManagement.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace carManagement.Services
{
    public class OrderService : IOrder
    {
        private readonly ApplicationDbContext _context;

        public OrderService(ApplicationDbContext context)
        {
            _context = context;
        }

        Order IOrder.PlaceOrder(int carId, int userId, int branchId, DateTime? startDate, DateTime? endDate, bool isPurchase)
        {
            var car = _context.Car.Find(carId);
            if (car == null)
            {
                throw new KeyNotFoundException("Car not found");
            }

            var inventory = _context.Inventory
                .FirstOrDefault(i => i.CarId == carId && i.BranchId == branchId);
            if (inventory == null)
            {
                throw new KeyNotFoundException("Inventory not found");
            }

            if (inventory.Quantity <= 0 && isPurchase)
            {
                throw new InvalidOperationException("Out of Stock");
            }

            OrderRequest order = new OrderRequest
            {
                CarId = carId,
                UserId = userId,
                StartDate = (DateTime)startDate,
                EndDate = (DateTime)endDate,
                IsPurchase = isPurchase
            };

            var branch = _context.Branch.Find(branchId);
            if (branch == null)
            {
                throw new KeyNotFoundException("Branch not found");
            }
            order.Branch = branch;

            double amount;
            if (isPurchase)
            {
                amount = car.PurchasePrice;
                inventory.Quantity -= 1;
            }
            else
            {
                if (!startDate.HasValue || !endDate.HasValue)
                {
                    throw new ArgumentException("Start date and end date must be provided for rentals.");
                }

                int rentalDays = (endDate.Value - startDate.Value).Days;
                if (rentalDays < 0)
                {
                    throw new ArgumentException("End date cannot be before start date.");
                }

                amount = rentalDays * car.PricePerDay;
            }

            _context.Order.Add(order);
            _context.Inventory.Update(inventory);
            _context.SaveChanges();

            var transaction = new Transaction
            {
                Order = order,
                Amount = amount,
                TransactionDate = DateTime.Now
            };
            _context.Transaction.Add(transaction);
            _context.SaveChanges();

            return order;
        }

        IEnumerable<Order> IOrder.GetAllOrders()
        {
            return _context.Order.ToList();
        }

        Order IOrder.GetOrderById(int id)
        {
            var order = _context.Order.Find(id);
            if (order == null)
            {
                throw new KeyNotFoundException("Order not found");
            }
            return order;
        }

        Order IOrder.UpdateOrder(int id, Order order)
        {
            if (id != order.OrderId)
            {
                throw new ArgumentException("Order ID mismatch");
            }

            if (!_context.Order.Any(o => o.OrderId == id))
            {
                throw new KeyNotFoundException("Order not found");
            }

            _context.Entry(order).State = Microsoft.EntityFrameworkCore.EntityState.Modified;
            _context.SaveChanges();
            return order;
        }

        void IOrder.DeleteOrder(int id)
        {
            var order = _context.Order.Find(id);
            if (order == null)
            {
                throw new KeyNotFoundException("Order not found");
            }

            _context.Order.Remove(order);
            _context.SaveChanges();
        }

        void IOrder.Save(OrderRequest orderRequest)
        {
            // Implement logic to save OrderRequest if needed
            throw new NotImplementedException();
        }

        bool IOrder.Exists(int id)
        {
            return _context.Order.Any(o => o.OrderId == id);
        }

        Order IOrder.Save(Order order)
        {
            throw new NotImplementedException();
        }
    }
}

File: CarMall_JavaBackend/src/main/java/com/org/carmallproject/Service/OrderService.java

package com.org.carmallproject.Service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.org.carmallproject.Entity.Branch;
import com.org.carmallproject.Entity.Car;
import com.org.carmallproject.Entity.Inventory;
import com.org.carmallproject.Entity.Order;
//import com.org.carmallproject.Entity.OrderRequest;
import com.org.carmallproject.Entity.Transaction;
import com.org.carmallproject.Repo.BranchRepository;
import com.org.carmallproject.Repo.CarRepository;
import com.org.carmallproject.Repo.InventoryRepository;
import com.org.carmallproject.Repo.OrderRepository;
import com.org.carmallproject.Repo.TransactionRepository;

import java.time.Duration;
import java.time.LocalDate;
import java.util.List;




@Service
public class OrderService {

    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private CarRepository carRepository;

    @Autowired
    private TransactionRepository transactionRepository;
    
    @Autowired
    private BranchRepository branchRepository;
    
    @Autowired
    private InventoryRepository inventoryRepository;
    
    public Order placeOrder(int carId, int userId, int branchId, LocalDate startDate, LocalDate endDate, boolean isPurchase) {
        // Fetch the car details
        Car car = carRepository.findById(carId)
                .orElseThrow(() -> new RuntimeException("Car not found"));

        
        Inventory inventory = inventoryRepository.findByCarIdAndBranchId(carId, branchId);
        
        if (inventory ==null ) {
        	throw new RuntimeException("Inventory not found");
        }
        
        if(inventory.getQuantity()<=0 && isPurchase) {
        	throw new RuntimeException("Out of Stock");
        }
        
        
        // Create and set up the order
        Order order = new Order();
        order.setCarId(carId);
        order.setUserId(userId);
        order.setStartDate(startDate);
        order.setEndDate(endDate);
        order.setPurchase(isPurchase);  // Directly use the parameter

        // Fetch and set the branch details
        Branch branch = branchRepository.findById(branchId)
                .orElseThrow(() -> new RuntimeException("Branch not found"));
        order.setBranch(branch);

        // Calculate the amount based on the order type (purchase or rental)
        double amount;
        if (isPurchase) {
            // Set amount to purchase price if it’s a purchase
            amount = car.getPurchasePrice();
            
            inventory.setQuantity(inventory.getQuantity()-1);
        } else {
            // Ensure start and end dates are provided for rental
            if (startDate == null || endDate == null) {
                throw new IllegalArgumentException("Start date and end date must be provided for rentals.");
            }
            int rentalDays = (int) Duration.between(startDate.atStartOfDay(), endDate.atStartOfDay()).toDays();
            if (rentalDays < 0) {
                throw new IllegalArgumentException("End date cannot be before start date.");
            }
            amount = rentalDays * car.getPricePerDay();
        }

        // Save the order to the repository
        orderRepository.save(order);
        inventoryRepository.save(inventory);
        // Create and save the transaction
        Transaction transaction = new Transaction();
        transaction.setOrder(order);  // Associate the entire Order object with the transaction
        transaction.setAmount(amount);  // Set the calculated amount
        transaction.setTransactionDate(LocalDate.now());  // Set the transaction date to today
        transactionRepository.save(transaction);

        return order;
    }

    public List<Order> getAllOrders() {
        return orderRepository.findAll();
    }

    public Order getOrderById(int id) {
        return orderRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Order not found"));
    }

    public Order updateOrder(int id, Order order) {
        if (!orderRepository.existsById(id)) {
            throw new RuntimeException("Order not found");
        }
        order.setOrderId(id);
        return orderRepository.save(order);
    }

    public void deleteOrder(int id) {
        orderRepository.deleteById(id);
    }
}
